### DISCUSSION TITLE
Custom 404 page not working, getting redirected to home page

### DISCUSSION THREAD

**Discussion by mageshyt** [vercel/next.js#49784] [App Router] [answered]
I am trying to set up a custom 404 page for my Next.js application, but when I enter a wrong path, I am getting redirected back to the home page instead of seeing the 404 page. Here are the details of the issue:

**Steps to reproduce:**
Enter a wrong path in the URL, such as "http://localhost:3000//this-page-does-not-exist"
Observe that instead of seeing the custom 404 page, I am redirected back to the home page.
**Expected behavior:**
I expect to see the custom 404 page I created (pages/404.js) when I enter a wrong path.

https://github.com/vercel/next.js/assets/70838644/0e9aa7b5-6838-4c59-b728-fc800a9c2176

**Comment by icyJoseph** [comment]
Hi,

I cannot reproduce this, using app dir, since you used the `App Router` label for your discussion. 

Could you give us more into how's this app working? 

- app directory tree structure
- do you have redirects/rewrites in next.config.js
- middleware?
- next.js version?

> I expect to see the custom 404 page I created (pages/404.js) when I enter a wrong path.

Now I am confused, you used the `App Router` label for this post, but you are using the pages directory? Are you using both? We need more info to be able to help you.

**Comment by mageshyt** [comment]
I am sorry for not proving more information 😅. i use page routing , 

 this is my version-1 build  https://exile-next.vercel.app/ in this build everything is working fine 

https://github.com/vercel/next.js/assets/70838644/79a03e17-6092-4a51-9170-012c4eb28597

**next version**
![image](https://github.com/vercel/next.js/assets/70838644/4a2aaf49-a594-4793-a298-6ff3f6610810)

**Next config**

![image](https://github.com/vercel/next.js/assets/70838644/0a1a41cf-ec77-43ea-840b-016dcd80b63a)

**my folder structure**
![image](https://github.com/vercel/next.js/assets/70838644/b1661d91-c7ba-41bc-be2a-9e912ae6764a)

**app.txt**
```tsx
function MyApp({ Component, pageProps }: AppProps) {
  return (
    // <SessionProvider session={pageProps}>
    <MantineProvider withGlobalStyles withNormalizeCSS>
      <Notifications position="top-right" limit={5} />
      <Layout>
        <LoginModal />
        <RegisterModal />
        {/* <LoginModal /> */}
        <Component {...pageProps} />
      </Layout>
    </MantineProvider>
    // </SessionProvider>
  );
}
```

**Reply by icyJoseph** [reply]
Alright, but this one does not have a `pages/404.tsx` page, and I guess the issue is that when you include it, it doesn't work, right?

**Reply by mageshyt** [reply]
### Thank you so much for your reply bro ,finally i found where the problem is 😮‍💨
```tsx
  useEffect(() => {
    if (searchInput.length >= 2) {
      router.push({
        pathname: "/search",
        query: { searchplace: searchInput },
      });
    }

    // if empty string
    if (searchInput.length === 0) {
      router.push({
        pathname: "/",
      });
    }
  }, [searchInput, router]);
```

 i guess  🤔 when the website refresh (or) move to different route the state's are getting re-initialized so that my useEffect is getting fired up .

https://github.com/vercel/next.js/assets/70838644/9707bfce-a699-4a92-a278-8b10e40412cd

**Reply by icyJoseph** [reply, marked as answer]
Hi,

Yeah, the `useRouter` result, changes when you `push`/`replace`, so you get a new router instance on every route change.


For this particular case, when you only use the `router.push` or `router.replace` methods, some folk use `import Router from 'next/router'`, and gives you access to a stable router instance, and you can skip it in the use-effect dependencies.

```jsx
import Router from 'next/router';

// later on inside your component...

  useEffect(() => {
    if (searchInput.length >= 2) {
      Router.push({
        pathname: "/search",
        query: { searchplace: searchInput },
      });
    }

    // if empty string
    if (searchInput.length === 0) {
      Router.push({
        pathname: "/",
      });
    }
  }, [searchInput]);
```

**Reply by mageshyt** [reply]
## thank you so much for your help 😄😄
